marimo/_utils/formatter.py

```diff
@@ -18,6 +20,51 @@ def format(self, codes: CellCodes) -> CellCodes:
         return codes
 
 
+class DefaultFormatter(Formatter):
+    """
+    Tries black, then ruff, then no formatting.
+    """
+
+    def format(self, codes: CellCodes) -> CellCodes:
+        if DependencyManager.has("black"):
+            return BlackFormatter(self.line_length).format(codes)
+        elif DependencyManager.has("ruff"):
+            return RuffFormatter(self.line_length).format(codes)
+        else:
+            LOGGER.warning(
+                "To enable code formatting, install black (pip install black) "
+                "or ruff (pip install ruff)"
+            )
+            return {}
+
+
+class RuffFormatter(Formatter):
+    def format(self, codes: CellCodes) -> CellCodes:
+        formatted_codes: CellCodes = {}
+        for key, code in codes.items():
+            try:
+                process = subprocess.run(
+                    [
+                        "ruff",
+                        "format",
+                        "--line-length",
+                        str(self.line_length),
+                        "-",
+                    ],
+                    input=code.encode(),
+                    capture_output=True,
+                    check=True,
+                )
+                formatted = process.stdout.decode()
+                formatted_codes[key] = formatted.strip()
```

Comment:
Before doing this, check to make sure the `process.returncode` is `0` (ie success)